src/audio/google_rtc_audio_processing.c

```diff
@@ -0,0 +1,385 @@
+						 sink_list);
+		buffer_lock(source_buffer, &flags);
+		if (source_buffer->source->comp.type == SOF_COMP_DEMUX)
+			cd->aec_reference = source_buffer;
+		else if (source_buffer->source->comp.type == SOF_COMP_DAI)
+			cd->raw_microphone = source_buffer;
+		buffer_unlock(source_buffer, flags);
+	}
+
+	cd->output = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+
+	switch (cd->output->stream.frame_fmt) {
+#if CONFIG_FORMAT_S16LE
+	case SOF_IPC_FRAME_S16_LE:
+		break;
+#endif /* CONFIG_FORMAT_S16LE */
+	default:
+		comp_err(dev, "unsupported data format: %d", cd->output->stream.frame_fmt);
+		return -EINVAL;
+	}
+
+	if (cd->output->stream.rate != GOOGLE_RTC_AUDIO_PROCESSING_SAMPLERATE) {
+		comp_err(dev, "unsupported samplerate: %d", cd->output->stream.rate);
+		return -EINVAL;
+	}
+
+	return comp_set_state(dev, COMP_TRIGGER_PREPARE);
+}
+
+static int google_rtc_audio_processing_reset(struct comp_dev *dev)
+{
+	comp_dbg(dev, "google_rtc_audio_processing_reset()");
+
+	return comp_set_state(dev, COMP_TRIGGER_RESET);
+}
+
+static int google_rtc_audio_processing_copy(struct comp_dev *dev)
+{
+	struct google_rtc_audio_processing_comp_data *cd = comp_get_drvdata(dev);
+
+	struct comp_copy_limits cl;
+	int frame;
+	uint32_t aec_reference_buff_frag;
+	uint32_t raw_microphone_buff_frag;
+	uint32_t output_buff_frag;
+	int16_t *src;
+	int16_t *dst;
+	uint32_t num_aec_reference_frames;
+	uint32_t num_aec_reference_bytes;
+	uint32_t flags;
+
+	buffer_lock(cd->aec_reference, &flags);
+	num_aec_reference_frames = audio_stream_get_avail_frames(&cd->aec_reference->stream);
+	num_aec_reference_bytes = audio_stream_get_avail_bytes(&cd->aec_reference->stream);
+	buffer_unlock(cd->aec_reference, flags);
+
+	buffer_invalidate(cd->aec_reference, num_aec_reference_bytes);
+
+	aec_reference_buff_frag = 0;
+	for (frame = 0; frame < num_aec_reference_frames; frame++) {
+		src = audio_stream_read_frag_s16(&cd->aec_reference->stream,
+										 aec_reference_buff_frag);
+		cd->aec_reference_buffer[cd->aec_reference_buffer_index] = *src;
+		++cd->aec_reference_buffer_index;
+
+		if (cd->aec_reference_buffer_index == cd->num_frames) {
+			GoogleRtcAudioProcessingAnalyzeRender_int16(cd->state,
+														cd->aec_reference_buffer);
+			cd->aec_reference_buffer_index = 0;
+		}
+		aec_reference_buff_frag += cd->aec_reference->stream.channels;
+	}
+	comp_update_buffer_consume(cd->aec_reference, num_aec_reference_bytes);
+
+	comp_get_copy_limits_with_lock(cd->raw_microphone, cd->output, &cl);
+	buffer_invalidate(cd->raw_microphone, cl.source_bytes);
+
+	raw_microphone_buff_frag = 0;
+	output_buff_frag = 0;
+	for (frame = 0; frame < cl.frames; frame++) {
+		src = audio_stream_read_frag_s16(&cd->raw_microphone->stream,
+										 raw_microphone_buff_frag);
+		cd->raw_mic_buffer[cd->raw_mic_buffer_index] = *src;
+		++cd->raw_mic_buffer_index;
+
+		dst = audio_stream_read_frag_s16(&cd->output->stream, output_buff_frag);
```

Comment:
@lkoenig Hi, is this intentional or mistake that sink buffer is written with reference to r_ptr with the use  function? Normally sink write is done with audio_stream_write_frag_s16() that references to w_ptr.

Reply:
Actually this is an error. Thanks for catching it. I'll fix it straight away.

Reply:
I am working on a patch for this component to deprecate read/write frag from SOF but it's good to have a correct reference, so please make the fix. Thanks!!

Reply:
https://github.com/thesofproject/sof/pull/5576 should correct it.

Reply:
Let me know if you want me to test the PR.

Reply:
@singalsu Let me know if you need help testing the read/write change.

Reply:
@lkoenig Nice, thank you! I'm able to run the mockup version in testbench and in a HDA generic device with a custom topology. But I really need your check and approve before it can be merged. I'll share in PR the MCPS saving that I see in my TGL-H.

Reply:
@lkoenig There's similar issue in Crossover. @andrula-song may be working on a fix. We are addressing rest of read/write frag together.

Reply:
Whelp, that is two I let slip through in my reviews, thanks @andrula-song and @singalsu !